### PR TITLE
feat(batch): Implement Lookup Join

### PR DESCRIPTION
**Pull request by Graphcalibur**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Summarize your change (**mandatory**)
    - Added the `LOOKUP JOIN` option for `LEFT` or `INNER` joins
    - The parsing, binding, and planning of Lookup Join has been added, however the implementation of `LookupJoinExecutor` is still incomplete as it is being blocked by the lack of `IndexScan`.
    - Note that `LogicalJoin` now has an extra `is_lookup_join` attribute. Similarly, all `JoinOperator` enums besides `CrossJoin` have an additional bool indicating whether they're lookup join or not (although Lookup Joins can only be used for LEFT and INNER joins).

## Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#2754

### CODE REVIEW

File: src/batch/src/executor/join/lookup_join.rs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+#[async_trait::async_trait]
+impl BoxedExecutorBuilder for LookupJoinExecutor {
+    async fn new_boxed_executor<C: BatchTaskContext>(
+        source: &ExecutorBuilder<C>,
+        mut inputs: Vec<BoxedExecutor>,
+    ) -> Result<BoxedExecutor> {
+        ensure!(
+            inputs.len() == 2,
+            "LookupJoinExeuctor should have 2 children!"
+        );
```
Comment (liurenjie1024) [inline src/batch/src/executor/join/lookup_join.rs:182]: No, lookup join has only one child.

File: proto/batch_plan.proto [hunk 2]
```diff
@@ -148,6 +148,14 @@ message TableFunctionNode {
   data.DataType return_type = 3;
 }
 
+message LookupJoinNode {
+  plan_common.JoinType join_type = 1;
+  repeated int32 left_key = 2;
+  repeated int32 right_key = 3;
```
Comment (liurenjie1024) [inline proto/batch_plan.proto:154]: The right side of `LookupJoinNode` should also contains table id and columns to construct `RowSeqScanNode`.

File: src/frontend/src/binder/relation/join.rs [hunk 3]
```diff
@@ -28,6 +28,7 @@ pub struct BoundJoin {
     pub left: Relation,
     pub right: Relation,
     pub cond: ExprImpl,
+    pub is_lookup_join: bool,
```
Comment (liurenjie1024) [inline src/frontend/src/binder/relation/join.rs:31]: The choice of lookup join should left to planner, e.g. converting logical join node to batch join node, rather in binder.

File: src/frontend/src/optimizer/plan_node/batch_lookup_join.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            eq_join_predicate,
+        }
+    }
+
+    fn derive_dist(left: &Distribution, right: &Distribution) -> Distribution {
+        match (left, right) {
+            (Distribution::Single, Distribution::Single) => Distribution::Single,
+            (_, _) => unreachable!(),
+        }
+    }
```
Comment (liurenjie1024) [inline src/frontend/src/optimizer/plan_node/batch_lookup_join.rs:59]: This distribution of `LookupJoin` should be same as probe side.

**Comment by liurenjie1024** [comment]
There maybe some misunderstanding about the lookup join implementation. The choice of join implementation(lookup join, hash join, nestedloop join) should left to planner, so we don't need to change parser and logical plan node. And to keep this PR small, I would suggest not to leave changes to frontend in next PR, and we focus on executor and plan node first.

**Comment by Graphcalibur** [comment]
The [design document](https://singularity-data.quip.com/vFwnA0WEl9Nt/Design-Lookup-Join), states that the user will be the one to decide whether to use LookupJoin or not. Are we moving the choice of which one to use from the user to the planner/optimizer instead? @liurenjie1024

**Comment by yuhao-su** [comment]
> There maybe some misunderstanding about the lookup join implementation. The choice of join implementation(lookup join, hash join, nestedloop join) should left to planner, so we don't need to change parser and logical plan node. And to keep this PR small, I would suggest not to leave changes to frontend in next PR, and we focus on executor and plan node first.

Currently, since we do not have CBO, I suggested using join hint or `SET` in this PR. But after a short discussion with @liurenjie1024 , it might be better to decide what join algorithm to use by using `SET` to 1) keep the syntax consistent with pgsql, 2) keep the query syntax consistent between batch and stream,3) we used `SET` to decide when to use delta join.

**Comment by yuhao-su** [comment]
> There maybe some misunderstanding about the lookup join implementation. The choice of join implementation(lookup join, hash join, nestedloop join) should left to planner, so we don't need to change parser and logical plan node. And to keep this PR small, I would suggest not to leave changes to frontend in next PR, and we focus on executor and plan node first.

Currently, since we do not have CBO, I suggested using join hint or `SET` in this PR. But after a short discussion with @liurenjie1024 , it might be better to decide what join algorithm to use by using `SET` to 1) keep the syntax consistent with pgsql, 2) keep the query syntax consistent between batch and stream,3) we used `SET` to decide when to use delta join.

File: src/batch/src/executor/join/lookup_join.rs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+impl LookupJoinExecutor {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        loop {
+            let cur_row = self.probe_side_source.get_current_row_ref();
+            if cur_row.is_some() {
+                // TODO: lookup rows on build side table using index scan, with cur_probe_row as
+                // input
```
Comment (liurenjie1024) [inline src/batch/src/executor/join/lookup_join.rs:71]: We don't need index scan here. The lookup join process is sending `RowSeqScanNode` to compute node to fetch result. It's quite similar to it: https://github.com/singularity-data/risingwave/blob/9abfc35b3c678cecef6260cf52fa6964cbe58a47/src/batch/src/execution/grpc_exchange.rs#L43

But you need to construct `RowSeqScanNode` using probe side data.

**Comment by Graphcalibur** [comment]
Closing this and creating another pull request at #3614 to make it cleaner

**Pull request closed without merging**